src/installer/corehost/cli/json_parser.cpp

```diff
@@ -140,12 +140,24 @@ bool json_parser_t::parse_file(const pal::string_t& path)
 
     if (bundle::info_t::is_single_file_bundle())
     {
+        // Due to in-situ parsing on Linux, 
+        //  * The json file is mapped as copy-on-write.
+        //  * The mapping cannot be immediately released, and will be unmapped by the json_parser destructor.
         m_bundle_data = bundle::info_t::config_t::map(path, m_bundle_location);
-        // The mapping will be unmapped by the json_parser destructor.
-        // The mapping cannot be immediately released due to in-situ parsing on Linux. 
 
         if (m_bundle_data != nullptr)
         {
+#if !defined(_WIN32)
+            // On Linux, the in-situ parser checks that the input stream is null-terminated
+            // once the root element is completely parsed.
+            // Ideally, the parser should check the size of the input stream, but it currently doesn't take the size into consideration.
+            // Instead of locally modifying RapidJson code, we simply add a null-terminator at the end of the stream.
+            // In the current implementation, it is safe to add the null-terminator because bundle::info_t::config_t::map() maps the
+            // enture single-file bundle, and there is content beyond all embeded files.
+            // This update is not performed on Windows in order to avoid an unnecessary copy-on-write.
+
+            m_bundle_data[m_bundle_location->size] = 0;
+#endif
             bool result = parse_json(m_bundle_data, m_bundle_location->size, path);
```

Comment:
Wouldn't it be better to alter `parse_json()` to use the `Parse()` [overloads that take a length](https://github.com/dotnet/runtime/blob/master/src/installer/corehost/cli/json/rapidjson/document.h#L2347-L2363)?

Reply:
I find this dangerous - what if there's real data right after the JSON in the single file. In that case we would be writing over it. If there's a way to tell the parser the length (like Leandro suggests) that would be a much cleaner solution.

Reply:
@Leandro: On Linux, the Json parser uses insitu parsing, which requires a null-terminated string. There is no variant of `ParseInsitu` that takes a size. 
https://github.com/dotnet/runtime/blob/master/src/installer/corehost/cli/json/rapidjson/document.h#L2296-L2314

@vitek-karas  The json files are mapped copy-on-write, so there's no risk in overwriting the content beyond the json file.

It is interesting to decide between the tradeoffs:
* Map the bundle copy-on-write and call `ParseInsitu()`. This will pretty-much copy all the pages.
* Map the bundle as read-only, and call `Parse()`. This will copy out strings to new buffers during parsing. 

So, I used the copy-on-write mapping approach because:
* The Rapid Json [documentation](https://rapidjson.org/md_doc_dom.html)  says that parsing incurs a large overhead because of copying individual strings to new places during parsing. Copying whole pages could be more efficient on modern hardware (thanks to write-buffers). Of course, we'll need to measure the difference to be sure. 
* It keeps the code somewhat simpler -- ex: no need to `ParseInsitu()` for normal apps, `Parse()` for single-file apps.

Reply:
OK - it will work... but I think it's worth a detailed comment explanation. It's not only that we map it as copy-on-write. It's also that we map it separately for each JSON we read (or rather we only read one JSON per mapping). Copy-on-write itself is not enough.
I wonder if there's some kind of assert or something we can do.

Honestly if we choose this solution I would actually prefer to modify the bundler to force-inject at least one byte of padding after the config files. Size wise it's trivial and it would make this always correct and robust (and we could make that padding a specific value and assert it in the hostpolicy).

Reply:
Just pass the kParseInSitu flag to the Parse() method that takes a length. ParseInsitu() is just a convenience method.

Reply:
In fact, I think that not passing the length is why the Windows version has the kParseStopWhenDoneFlag... hindsight is 20/20. Maybe `parse_json()` should be something like this?

```c
    constexpr auto flags = rapidjson::ParseFlag::kParseStopWhenDoneFlag
        | rapidjson::ParseFlag::kParseCommentsFlag
#ifndef _WIN32
        | rapidjson::ParseFlag::kParseInSituFlag
# // _WIN32
    ;

    m_document.Parse<flags>(data, length);
```

Reply:
@lpereira, the `Parse()` variants that take a size arguments use `memorystream`  and are not compatible with `in-situ` parsing. Here's the [check](https://github.com/dotnet/runtime/blob/master/src/installer/corehost/cli/json/rapidjson/document.h#L2349) that precludes using `Parse()` with `kParseInsituFlag`.

Reply:
However, from your comments, I noticed that `kParseStopWhenDoneFlag` flag, which serves this purpose on Windows. So, I've updated the PR to use `kParseStopWhenDoneFlag` on all architectures.

Like @vitek-karas  suggested above, we could just as easily add the null terminator from the bundle generator. However, this approach of using `kParseStopWhenDoneFlag`  is simpler (less coupling between bundler and hosts) and makes the parser behavior more consistent across architectures.

I've updated the PR notes above wrt to the new fix. Thanks.